text/0000-default-object-bound.md

```diff
@@ -0,0 +1,312 @@
+mut (SomeTrait+'b)`, respectively. In the case of a shared reference
+`&'a SomeTrait`, this difference is basically irrelevant, as the
+lifetime bound can always be approximated to be shorter when needed.
+
+In the case a mutable reference `&'a mut SomeTrait`, however, using
+two lifetime variables is *in principle* a more general expansion. The
+reason has to do with "variance" -- specifically, because the proposed
+expansion places the `'a` lifetime qualifier in the reference of a
+mutable reference, the compiler will be unable to allow `'a` to be
+approximated with a shorter lifetime. You may have experienced this if
+you have types like `&'a mut &'a mut Foo`; the compiler is also forced
+to be conservative about the lifetime `'a` in that scenario.
+
+However, in the specific case of object types, this concern is
+ameliorated by the existing object coercions. These coercions permit
+`&'a mut (SomeTrait+'a)` to be coerced to `&'b mut (SomeTrait+'c)`
+where `'a : 'b` and `'a : 'c`. The reason that this is legal is
+because unsized types (like object types) cannot be assigned, thus
+sidestepping the variance concerns. This means that programs like the
+following compile successfully (though you will find that you get
+errors if you replace the object type `(Counter+'a)` with the
+underlying type `&'a mut u32`):
+
+```rust
+#![allow(unused_variables)]
+#![allow(dead_code)]
+
+trait Counter {
+    fn inc_and_get(&mut self) -> u32;
+}
+
+impl<'a> Counter for &'a mut u32 {
+    fn inc_and_get(&mut self) -> u32 {
+        **self += 1;
+        **self
+    }
+}
+
+fn foo<'a>(x: &'a u32, y: &'a mut (Counter+'a)) {
+}
+
+fn bar<'a>(x: &'a mut (Counter+'a)) {
+    let value = 2_u32;
+    foo(&value, x)
+}
+
+fn main() {
+}
+```
+
+This may seem surprising, but it's a reflection of the fact that
+object types give the user less power than if the user had direct
+access to the underlying data; the user is confined to accessing the
+underlying data through a known interface.
+
+# Drawbacks
+
+**A. Breaking change.** This change has the potential to break some
+existing code, though given the statistics gathered we believe the
+effect will be minimal (in particular, defaults are only permitted in
+fn signatures today, so in most existing code explicit lifetime bounds
+are used).
+
+**B. Hidden references get the wrong default.** If you have a struct
+with a reference hidden within one of the fields, the defaults that
+result might not be what you desire. Consider the following:
+
+```rust
+struct Ref<'a, T: ?Sized + 'a> {
+    reference: &'a T
+}
+```
+
+Now if I have a type `Ref<'a, SomeTrait>`, I might prefer for this
+type to expand to `Ref<'a, SomeTrait+'a>`. However, given the rules as
+proposed here, the result would be `Ref<'a, SomeTrait+'static>`. If
+this becomes a serious annoyance, in the future we could add the
+option for declaring on the struct `Ref` that objects which appear in
+type `T` should be defaulted with the lifetime `'a`.
```

Comment:
It seems to me that this change would not be backwards compatible (it would change the bound from 'static to 'a).

So I think this needs to be implemented now or it must be an error to not specify a bound in any case where an additional rule is to be implemented in the future.